solutions/gold/rmi-17-Hangman2.mdx

```diff
@@ -0,0 +1,147 @@
+---
+id: rmi-17-Hangman2
+source: RMI 2017 Day 1
+title: Hangman 2
+author: Justin Ji
+---
+
+<Spoiler title = "Hint">
+
+Consider breaking the problem into two cases: $n < k$ and $n \geq k$.
+
+</Spoiler>
+
+<Spoiler title = "Explanation">
+
+First, note that $min(n, k) \leq \sqrt{nk}$. This will be important for
+analyzing the time complexity of our solution. Now, let's consider the two
+cases given in the hint.
+
+### Case 1: $n < k$
+Because $n < \sqrt{nk}$, we can brute force each pair of strings and check if
+they differ by no more than two characters. This takes $\mathcal{O}(N^2K)$ time,
+which is equivalent to $\mathcal{O}(NK\sqrt{NK})$ time.
+
+### Case 2: $n \geq k$
+Let's compare two strings. Note how allowing at most two characters to be
+different is equivalent to ignoring two characters in both strings and checking
+for equality. Thus, we can brute force on every pair of characters that we are
+ignoring, and then check for equality between strings.
+
+Let the pair of characters we are brute forcing be $(i, j)$. Then,
+we have the following intervals that need to be identical in both strings:
+
+$[0, i - 1], [i + 1, j - 1], [j + 1, k - 1]$
+
+Now, if we want to compare our two strings, then we just need the substrings to
+be equal. This can be checked using string hashing.
+
+For each string in our list, we need to find another string that has the same
+hashes. This is best done by sorting all the tuples of hashes, and then checking
+adjacent items and seeing if the hashes are equal.
+
+Overall, this takes $\mathcal{O}(NK^2\log({N}))$ time. With a bit more analysis,
+you can reach the time complexity of $\mathcal{O}(NK\sqrt{NK}\log(\sqrt{NK}))$.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(NK\sqrt{NK}\log(\sqrt{NK}))$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
+
+class HashedString {
```

Comment:
codesnip
also this is straight from the module right

Reply:
oh yeah i should do that

uhh, i made some minor edits to it. I can just yoink the one from the module tho (will do later)

Reply:
ok done, there are some very minor tweaks to make C++ stop complaining but other than that its the same

Reply:
wait what's c++ complaining about

Reply:
1. it was complaining about `vector<HashedString>`, so I created an empty initializer
2. C++ compiler was complaining about comparing `int` to `size_t` in one of the loops, so I casted it